activation/nipost_test.go

```diff
@@ -807,6 +807,299 @@ func TestNIPSTBuilder_PoetUnstable(t *testing.T) {
 	})
 }
 
+// TestNIPoSTBuilder_PoETConfigChange checks if
+// it properly detects added/deleted PoET services and re-registers if needed.
+func TestNIPoSTBuilder_PoETConfigChange(t *testing.T) {
+	t.Parallel()
+
+	sig, err := signing.NewEdSigner()
+	require.NoError(t, err)
+
+	challenge := types.NIPostChallenge{
+		PublishEpoch: postGenesisEpoch + 2,
+	}
+
+	challengeHash := wire.NIPostChallengeToWireV1(&challenge).Hash()
+
+	const (
+		poetProverAddr  = "http://localhost:9999"
+		poetProverAddr2 = "http://localhost:9988"
+	)
+
+	t.Run("1 poet deleted BEFORE round started -> continue with submitted registration", func(t *testing.T) {
+		db := localsql.InMemory()
+		ctrl := gomock.NewController(t)
+
+		poet := NewMockPoetService(ctrl)
+		poet.EXPECT().
+			Submit(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).
+			AnyTimes().
+			Return(&types.PoetRound{}, nil)
+		poet.EXPECT().Address().AnyTimes().Return(poetProverAddr).AnyTimes()
+
+		// successfully registered to 2 poets
+		err = nipost.AddPoetRegistration(db, sig.NodeID(), nipost.PoETRegistration{
+			ChallengeHash: challengeHash,
+			Address:       poetProverAddr,
+			RoundID:       "1",
+			RoundEnd:      time.Now().Add(1 * time.Second),
+		})
+		require.NoError(t, err)
+
+		err = nipost.AddPoetRegistration(db, sig.NodeID(), nipost.PoETRegistration{
+			ChallengeHash: challengeHash,
+			Address:       poetProverAddr2,
+			RoundID:       "1",
+			RoundEnd:      time.Now().Add(1 * time.Second),
+		})
+
+		nb, err := NewNIPostBuilder(
+			db,
+			nil,
+			zaptest.NewLogger(t),
+			PoetConfig{},
+			nil,
+			nil,
+			WithPoetServices(poet), // add only 1 poet prover
+		)
+		require.NoError(t, err)
+
+		existingRegistrations, err := nb.submitPoetChallenges(
+			context.Background(),
+			sig,
+			time.Now().Add(10*time.Second),
+			time.Now().Add(5*time.Second),
+			challengeHash.Bytes())
+
+		require.NoError(t, err)
+		require.Len(t, existingRegistrations, 1)
+		require.Equal(t, poetProverAddr, existingRegistrations[0].Address)
```

Comment:
I don't think the node should delete any registration by itself even if the config doesn't contain that PoET.

Reply:
It doesn't delete registration: it just returns only registrations, using which you can fetch proof from poets, which are configured. 

So if you have submitted registrations to poets A, B C, than removed poet A and restarted node, submitPoetChallenges will return registrations only for poet B and C. Since poet A is not configured anymore, there is no sense return registration for it.

Reply:
See my second comment: 

> Since they aren't actually deleted instead it would be good to at least log a warning.